pkg/config/decode.go

```diff
@@ -0,0 +1,120 @@
+	// Such cases are most likely due to a broken TOML configuration or invalid env variable value.
+	defer func() {
+		if r := recover(); r != nil {
+			err = r.(error)
+		}
+	}()
+
+	if lookupEnv == nil {
+		lookupEnv = os.LookupEnv
+	}
+	d := &decoder{
+		tree:      tree,
+		lookupEnv: lookupEnv,
+	}
+
+	conf = DefaultConfig()
+	d.unmarshal(decodeContext{
+		val: reflect.ValueOf(&conf),
+		typ: reflect.TypeOf(&conf),
+	})
+
+	conf.tree = tree
+	return conf, nil
+}
+
+type decoder struct {
+	tree      *toml.Tree
+	lookupEnv func(key string) (string, bool)
+}
+
+type decodeContext struct {
+	val      reflect.Value
+	typ      reflect.Type
+	tomlPath []string
+	envVar   string
+}
+
+func (d *decoder) unmarshal(ctx decodeContext) {
+	// If the current type is a pointer, get the actual type/value behind it.
+	if ctx.typ.Kind() == reflect.Ptr {
+		ctx.typ = ctx.typ.Elem()
+		ctx.val = ctx.val.Elem()
+	}
+
+	switch ctx.typ.Kind() {
+	case reflect.Struct:
+		// The current type is a struct, iterate through its fields.
+		for i := 0; i < ctx.typ.NumField(); i++ {
+			field := ctx.typ.Field(i)
+
+			// Lookup for the "toml" tag and skip this field if not present.
+			tomlKey, ok := field.Tag.Lookup("toml")
+			if !ok {
+				continue
+			}
+
+			// Generate a new decodeContext from the struct field and continue unmarshaling.
+			fieldVal := ctx.val.FieldByName(field.Name)
+			d.unmarshal(decodeContext{
+				val:      fieldVal,
+				typ:      fieldVal.Type(),
+				tomlPath: append(ctx.tomlPath, tomlKey),
+				envVar:   field.Tag.Get("env"),
+			})
+		}
+	case reflect.String, reflect.Int, reflect.Bool:
+		// Check if the current type supports env var overriding and look-it up if needed.
+		if ctx.envVar != "" {
+			if envValStr, ok := d.lookupEnv(ctx.envVar); ok {
+				if envVal, err := convertStringTo(envValStr, ctx.typ); err == nil {
+					ctx.val.Set(reflect.ValueOf(envVal))
+					return
+				}
+			}
+		}
+
+		// Fallback to reading from the TOML tree.
+		tomlNode := d.tree.GetPath(ctx.tomlPath)
+		if tomlNode != nil {
+			tomlVal := reflect.ValueOf(tomlNode)
+			if tomlVal.Type().ConvertibleTo(ctx.typ) {
+				ctx.val.Set(tomlVal.Convert(ctx.typ))
+			}
+		}
+	default:
+		// This should not happen as we're reflecting a well-known struct Config. A panic here means
+		// that a new field type has been introduced in the Config, thus requires changes in the decoder.
+		panic(fmt.Errorf("unsupported config type \"%s\"", ctx.typ))
+	}
+}
+
+// convertStringTo converts a string to a given type. It returns an error if the conversion can't be done.
+func convertStringTo(str string, targetType reflect.Type) (interface{}, error) {
+	switch targetType.Kind() {
+	case reflect.String:
+		return str, nil
+	case reflect.Bool:
+		return strconv.ParseBool(str)
+	case reflect.Int:
+		return strconv.Atoi(str)
+	default:
+		return nil, fmt.Errorf("couldn't convert string \"%s\" to \"%s\"", str, targetType)
```

Comment:
According to the Go codebase, it's OK to have contractions in error messages (e.g. https://github.com/golang/go/blob/964639cc338db650ccadeafb7424bc8ebb2c0f6c/misc/cgo/testsigfwd/main.go#L107) :+1:.